### ISSUE TITLE
Allow multibranch conditionals

### ISSUE THREAD

**Issue by pvdrz** [pvdrz/pijama_legacy#34]
Pijama doesn't have an `else if`. Life would be easier (except for the maintainers) if it had one.

We need to
- [x] figure out which syntax to use. (It seems we are going with `if v do w elif x do y else z end`)
- [x] modify the parser to support such syntax.
- [x] add the new node variant to the AST.
- [x] lower the new node variant as a chain of nested conditionals.
- [x] add tests for parsing.
- [x] update the tests and examples to use the new syntax.

**Comment by pvdrz** [comment]
I have three candidates in mind for the syntax:

- Traditional else if:
```
if x == 0 do
    0
else if x < 0 do
    -1
else
    1
end
```
- Python's `elif`:
```
if x == 0 do
    0
elif x < 0 do
    -1
else
    1
end
```
- A new `cond` construct (which goes in the pattern matching direction)
```
cond do 
    x == 0 -> 0,
    x < 0 -> -1,
    true -> 1,
end
```

**Comment by DarkDrek** [comment]
Python's `elif` seems to be the most straightforward and easiest to implement/parse from the current system.
But I like the `cond` idea would it remove/replace the traditional `if`? How would a longer body look like?
```
cond do 
    x == 0 -> do
        1
    end
    true -> do
        -1
    end
end
```

**Comment by pvdrz** [comment]
I wouldn't remove the regular `if do else end` syntax to replace it with the `cond` one. with the `->` and the `,` acting as block delimiters we can just do multi-line statements without doing fancy parsing:
```
cond do 
    x == 0 ->
        res = 1
        1,
    true -> -1,
end
```
The only thing i don't like about this syntax is that it doesn't look uniform at all, `do` and `end` are the usual block delimiters. However, we need a syntax for pattern matching eventually.

**Comment by DarkDrek** [comment]
Than I would make the pattern matching syntax another ticket and keep multibranch conditionals in line with the current syntax.
Can we do?
```
if x == 0 do
    0
if x < 0 do
    -1
else
    1
end
```

**Comment by pvdrz** [comment]
> Than I would make the pattern matching syntax another ticket and keep multibranch conditionals in line with the current syntax.
> Can we do?
> 
> ```
> if x == 0 do
>     0
> if x < 0 do
>     -1
> else
>     1
> end
> ```

What do you mean by "can we do?"? 

If you mean if it is valid code right now. It isn't, but it almost is. 

If you mean if we can consider that syntax to be our `else if`. My problem with it is that it is additional `else` away to be a completely different program (indentation is not considered during parsing):
```
if x == 0 do
    0
    if x < 0 do
        -1
    else
        1
    end
else 
    // more stuff
end
```
which means we'd need to do longer lookaheads to decide what the chain of conditionals means. I think for now the `elif` is the best option in terms of parsing complexity.

**Comment by seanchen1991** [comment]
Pattern matching definitely sounds like a good thing to add in the future. If we opt for an `elif` syntax in the short term, will it all need to be ripped out in the future if/when we decide to replace it with `cond`?

**Comment by pvdrz** [comment]
I think we can keep `elif` even if we decide to go into the `cond` direction as `elif` is a more common construct.

**Comment by seanchen1991** [comment]
Will `if`-`elif` in Pijama work the same as in a Lisp-like language where only one `elif` statement is allowed for an `if` statement? Or do we want to allow multiple `elif` statements for an `if`?

**Comment by pvdrz** [comment]
I think we can allow as much `elif`s as we want. However we might need to change the IRs to something more manageable. Nesting `Term::Cond`s might introduce a lot of indirections. So I think something like:
```rust
struct Branch<'a> 
     cond: Term<'a>,
     body: Term<'a>,
}

enum Term<'a> {
    ... 
    Cond(Vec<Branch<'a>>, Box<Term<'a>>),
}
```
Where the branches represents all the possible `if  cond do body` or `elif cond do body` and the second term in `Cond` represents the `else term end` part.

**Comment by seanchen1991** [comment]
I'd like to take a stab at implementing this 🙂

**Comment by pvdrz** [comment]
Great! don't hesitate to ask if you have any questions me :)

**Comment by pvdrz** [comment]
@seanchen1991 did you already start working on this? I'm asking because I want to merge https://github.com/christianpoveda/pijama/pull/49 but I don't want to cause any merge issues for you.

**Comment by seanchen1991** [comment]
I'm not far enough along that I would mind restarting; go ahead and merge it in.

**Comment by pvdrz** [comment]
Oh that's good to know. Thanks for the heads up :)

**Comment by pvdrz** [comment]
#74 has taken care of most of the tasks posted here. However, we still need to figure out parsing. For that I think we should introduce a new combinator able to parse branches inside the `parser::node::cond` module. We have two kinds of branches now `if <block> do <block>` and `elif <block> do <block>`. We could write a single combinator
```rust
fn branch(kwrd: &str) -> impl Fn(Span) -> IResult<Located<Branch>> { ... }
```
where `branch("if")` is able to parse the first kind of branches and `branch("elif")` the second kind.

After writing that combinator we can rewrite the `cond` parser to be something like
```rust
pub fn cond(input: Span) -> IResult<Located<Node>> {
    map(
        tuple((
            position,
            branch("if"),
            many0(branch("elif")),
            else_block,
            preceded(keyword("end"), position),
        )),
        move |(sp1, if_branch, branches, else_block, sp2)| {
            Located::new(
                Node::Cond(if_branch, branches, else_block),
                Location::from(sp1) + Location::from(sp2),
            )
        },
    )(input)
}
```
cc @seanchen1991

**Comment by seanchen1991** [comment]
My PR is already pretty sizable at this point. I'll update current `eval` tests to utilize `elif` statements in a separate PR.

**Comment by pvdrz** [comment]
closed via #74
